src/app/tests/suites/include/ConstraintsChecker.h

```diff
@@ -309,12 +338,41 @@ class ConstraintsChecker
         return true;
     }
 
-    template <typename T, typename U, std::enable_if_t<std::is_enum<T>::value && !std::is_pointer<U>::value, int> = 0>
+    template <typename T, typename U,
+              std::enable_if_t<std::is_enum<T>::value && !std::is_enum<U>::value && !std::is_pointer<U>::value, int> = 0>
     bool CheckConstraintMaxValue(const char * itemName, T current, U expected)
     {
         return CheckConstraintMaxValue(itemName, chip::to_underlying(current), expected);
     }
 
+    template <typename T, typename U,
+              std::enable_if_t<std::is_enum<T>::value && std::is_enum<U>::value && std::is_convertible<U, int>::value, int> = 0>
+    bool CheckConstraintMaxValue(const char * itemName, T current, U expected)
+    {
+        return CheckConstraintMaxValue(itemName, chip::to_underlying(current), chip::to_underlying(expected));
+    }
+
+    template <typename T, typename U,
+              std::enable_if_t<std::is_enum<T>::value && std::is_enum<U>::value && !std::is_convertible<U, int>::value, int> = 0>
```

Comment:
I am starting to wonder whether this would be simpler if we had a `AsRawInteger` template that basically took a value and had two enable_if overloads: if the type is enum, return `to_underlying(value)`, and if the type is not enum just return the value.  Then instead of having to have 4 cases for all combinations of (first thing enum, second thing enum), we could just have one case that calls `AsRawInteger` on both the current and expected.  Thoughts?